linera-service/src/util.rs

```diff
@@ -1,19 +1,78 @@
 // Copyright (c) Zefchain Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use anyhow::{bail, Result};
+use anyhow::{anyhow, bail, Result};
 use std::path::PathBuf;
+use tokio::process::Command;
+use tracing::{debug, error};
 
-pub fn resolve_cargo_binary(bin_name: &'static str) -> Result<PathBuf> {
-    cargo_home().map(|cargo_home| cargo_home.join("bin").join(bin_name))
-}
+/// Attempt to resolve the path and test the version of the given binary against our
+/// package version. This is meant for binaries of the Linera repository.
+pub async fn resolve_binary(name: &'static str, package: &'static str) -> Result<PathBuf> {
+    debug!("Resolving binary {name} based on the current binary path.");
+    let mut current_binary_path = PathBuf::from(
+        std::env::args()
+            .next()
+            .expect("args should start with the current binary path"),
+    );
+    current_binary_path.pop();
+
+    #[cfg(any(test, feature = "test"))]
+    // Test binaries are typically in target/debug/deps while crate binaries are in target/debug
+    // (same thing for target/release).
+    {
+        let components = current_binary_path
+            .components()
+            .map(std::path::Component::as_os_str)
+            .collect::<Vec<_>>();
+        let len = components.len();
+        if len >= 3
+            && components[len - 3] == "target"
+            && (components[len - 2] == "debug" || components[len - 2] == "release")
+            && components[len - 1] == "deps"
+        {
+            current_binary_path.pop();
+        }
+    }
 
-fn cargo_home() -> Result<PathBuf> {
-    if let Ok(cargo_home) = std::env::var("CARGO_HOME") {
-        Ok(PathBuf::from(cargo_home))
-    } else if let Some(home) = dirs::home_dir() {
-        Ok(home.join(".cargo"))
-    } else {
-        bail!("could not find CARGO_HOME directory, please specify it explicitly")
+    let path = current_binary_path.join(name);
+    let version = env!("CARGO_PKG_VERSION");
+    if !path.exists() {
+        error!(
+            "Cannot find a binary {name} in the directory {}. \
+             Consider using `cargo install {package}` or `cargo build -p {package}`",
+            current_binary_path.display()
+        );
+        bail!("Failed to resolve binary {name}");
     }
+
+    // Quick version check.
+    let version_message = Command::new(&path)
+        .arg("--version")
+        .output()
+        .await
+        .map_err(|e| {
+            anyhow!(
+            "Failed to execute and retrieve version from the binary {name} in directory {}: {e}",
+            current_binary_path.display())
+        })?
+        .stdout;
+    let found_version = String::from_utf8_lossy(&version_message)
+        .trim()
+        .split(' ')
+        .last()
+        .ok_or(anyhow!(
```

Comment:
(Doesn't make much of a difference here, but consider using `ok_or_else` to avoid allocating the string.)